File: codetandem/state.py
"""Project state management and codetandem.state.json generation."""

import json
from datetime import datetime, timezone
from pathlib import Path
from typing import Dict, Union

from codetandem.modules import load_modules
from codetandem.scanner import FileTree, get_file_tree


def generate_initial_state(
    modules_path: Union[str, Path],
    project_path: Union[str, Path],
    output_path: Union[str, Path],
    initial_skill_score: float = 0.0,
) -> Dict:
    """
    Generate initial codetandem.state.json file.

    Args:
        modules_path: Path to modules.json file
        project_path: Path to the project directory to scan
        output_path: Path where state.json should be written
        initial_skill_score: Initial skill score (default 0.0)

    Returns:
        Dictionary containing the state data that was written

    Raises:
        FileNotFoundError: If modules.json or project directory doesn't exist
        ValueError: If modules.json has no modules
    """
    # Load modules
    modules = load_modules(modules_path)

    if not modules:
        raise ValueError("No modules found in modules.json")

    # Get the first module
    first_module = modules[0]

    # Scan project directory
    file_tree = get_file_tree(project_path)

    # Generate initial state
    state_data = {
        "version": "1.0",
        "created_at": datetime.now(timezone.utc).isoformat(),
        "updated_at": datetime.now(timezone.utc).isoformat(),
        "current_module_id": first_module.id,
        "skill_scores": {first_module.id: initial_skill_score},
        "completed_modules": [],
        "project_tree": file_tree,
        "metadata": {
            "total_modules": len(modules),
            "project_path": str(Path(project_path).resolve()),
        },
    }

    # Write to file
    output_path = Path(output_path)
    output_path.parent.mkdir(parents=True, exist_ok=True)

    with open(output_path, "w", encoding="utf-8") as f:
        json.dump(state_data, f, indent=2, ensure_ascii=False)

    return state_data


def load_state(state_path: Union[str, Path]) -> Dict:
    """
    Load state from codetandem.state.json file.

    Args:
        state_path: Path to state.json file

    Returns:
        Dictionary containing the state data

    Raises:
        FileNotFoundError: If state.json doesn't exist
        ValueError: If state.json is invalid
    """
    state_path = Path(state_path)

    if not state_path.exists():
        raise FileNotFoundError(f"State file not found: {state_path}")

    with open(state_path, "r", encoding="utf-8") as f:
        data = json.load(f)

    # Validate required fields
    required_fields = ["current_module_id", "skill_scores", "completed_modules"]
    for field in required_fields:
        if field not in data:
            raise ValueError(f"Invalid state.json: missing '{field}' field")

    return data


def update_state(
    state_path: Union[str, Path],
    current_module_id: str = None,
    skill_score: float = None,
    completed_module_id: str = None,
    project_tree: FileTree = None,
    hint_count: int = None,
    difficulty_override: str = None,
    assessment_pending: bool = None,
) -> Dict:
    """
    Update an existing state file.

    Args:
        state_path: Path to state.json file
        current_module_id: New current module ID
        skill_score: Update skill score for current module
        completed_module_id: Module ID to mark as completed
        project_tree: Updated project tree

    Returns:
        Updated state dictionary

    Raises:
        FileNotFoundError: If state.json doesn't exist
    """
    state = load_state(state_path)

    # Update current module
    if current_module_id is not None:
        state["current_module_id"] = current_module_id

    # Update skill score
    if skill_score is not None:
        current_id = state["current_module_id"]
        state["skill_scores"][current_id] = skill_score

    # Mark module as completed
    if completed_module_id is not None:
        if completed_module_id not in state["completed_modules"]:
            state["completed_modules"].append(completed_module_id)

    # Update project tree
    if project_tree is not None:
        state["project_tree"] = project_tree

    # Update hint count
    if hint_count is not None:
        current_id = state["current_module_id"]
        if "hints" not in state:
            state["hints"] = {}
        state["hints"][current_id] = hint_count

    # Update difficulty override
    if difficulty_override is not None:
        state["difficulty_override"] = difficulty_override

    # Update assessment pending status
    if assessment_pending is not None:
        state["assessment_pending"] = assessment_pending

    # Update timestamp
    state["updated_at"] = datetime.now(timezone.utc).isoformat()

    # Write back to file
    with open(state_path, "w", encoding="utf-8") as f:
        json.dump(state, f, indent=2, ensure_ascii=False)

    return state


def get_current_module_id(state: Dict) -> str:
    """Get the current module ID from state."""
    return state["current_module_id"]


def get_skill_score(state: Dict, module_id: str) -> float:
    """Get the skill score for a module."""
    return state["skill_scores"].get(module_id, 0.0)


def is_module_completed(state: Dict, module_id: str) -> bool:
    """Check if a module is completed."""
    return module_id in state["completed_modules"]


def get_hint_count(state: Dict, module_id: str) -> int:
    """Get the hint count for a module."""
    return state.get("hints", {}).get(module_id, 0)


def increment_skill_score(
    state_path: Union[str, Path], module_id: str, points: float
) -> float:
    """
    Increment the skill score for a module.

    Args:
        state_path: Path to state.json file
        module_id: Module ID to update
        points: Points to add to current score

    Returns:
        New skill score
    """
    state = load_state(state_path)

    current_score = state["skill_scores"].get(module_id, 0.0)
    new_score = current_score + points

    # Update the score
    state["skill_scores"][module_id] = new_score
    state["updated_at"] = datetime.now(timezone.utc).isoformat()

    # Write back to file
    with open(state_path, "w", encoding="utf-8") as f:
        json.dump(state, f, indent=2, ensure_ascii=False)

    return new_score

File: src/utils/state.ts
/**
 * Project state management
 */

import { readFile, writeFile, mkdir } from 'fs/promises';
import { dirname } from 'path';
import type { StateData, UpdateStateOptions, FileTree } from '../types/state.js';

/**
 * Load state from codetandem.state.json file
 */
export async function loadState(statePath: string): Promise<StateData> {
  try {
    const content = await readFile(statePath, 'utf-8');
    const data = JSON.parse(content) as StateData;

    // Validate required fields
    const requiredFields: (keyof StateData)[] = [
      'currentModuleId',
      'skillScores',
      'completedModules',
    ];

    for (const field of requiredFields) {
      if (!(field in data)) {
        throw new Error(`Invalid state.json: missing '${field}' field`);
      }
    }

    return data;
  } catch (error) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    if ((error as any)?.code === 'ENOENT') {
      throw new Error(`State file not found: ${statePath}`);
    }
    throw error;
  }
}

/**
 * Save state to file
 */
async function saveState(statePath: string, state: StateData): Promise<void> {
  // Ensure directory exists
  await mkdir(dirname(statePath), { recursive: true });

  // Write state file
  await writeFile(statePath, JSON.stringify(state, null, 2), 'utf-8');
}

/**
 * Generate initial state file
 */
export async function generateInitialState(options: {
  modulesPath: string;
  projectPath: string;
  outputPath: string;
  initialSkillScore?: number;
  firstModuleId: string;
  totalModules: number;
  projectTree?: FileTree;
}): Promise<StateData> {
  const {
    outputPath,
    initialSkillScore = 0.0,
    firstModuleId,
    totalModules,
    projectTree,
    projectPath,
  } = options;

  const now = new Date().toISOString();

  const stateData: StateData = {
    version: '1.0',
    createdAt: now,
    updatedAt: now,
    currentModuleId: firstModuleId,
    skillScores: { [firstModuleId]: initialSkillScore },
    completedModules: [],
    projectTree,
    metadata: {
      totalModules,
      projectPath,
    },
  };

  await saveState(outputPath, stateData);
  return stateData;
}

/**
 * Update an existing state file
 */
export async function updateState(
  statePath: string,
  options: UpdateStateOptions
): Promise<StateData> {
  const state = await loadState(statePath);
  const currentId = state.currentModuleId;

  // Initialize moduleProgress if needed
  if (!state.moduleProgress) {
    state.moduleProgress = {};
  }
  if (!state.moduleProgress[currentId]) {
    state.moduleProgress[currentId] = {
      attempts: 0,
      objectivesCompleted: [],
      hintsUsed: 0,
      solutionsUsed: 0,
      bestScore: 0,
    };
  }

  // Update current module
  if (options.currentModuleId !== undefined) {
    state.currentModuleId = options.currentModuleId;
  }

  // Update skill score for current module
  if (options.skillScore !== undefined) {
    state.skillScores[currentId] = options.skillScore;
    // Update best score
    if (options.skillScore > state.moduleProgress[currentId]!.bestScore) {
      state.moduleProgress[currentId]!.bestScore = options.skillScore;
    }
  }

  // Record objective completion
  if (options.objectiveCompletion !== undefined) {
    const progress = state.moduleProgress[currentId]!;

    // Check if objective already completed (by objectiveId or todoId)
    const alreadyCompleted = progress.objectivesCompleted.some(
      (obj) =>
        obj.objectiveId === options.objectiveCompletion!.objectiveId ||
        (obj.todoId && obj.todoId === options.objectiveCompletion!.todoId)
    );

    if (!alreadyCompleted) {
      progress.objectivesCompleted.push(options.objectiveCompletion);
    }
  }

  // Increment hints used
  if (options.incrementHints) {
    state.moduleProgress[currentId]!.hintsUsed += 1;
  }

  // Increment solutions used
  if (options.incrementSolutions) {
    state.moduleProgress[currentId]!.solutionsUsed += 1;
  }

  // Mark module as completed
  if (options.completedModuleId !== undefined) {
    if (!state.completedModules.includes(options.completedModuleId)) {
      state.completedModules.push(options.completedModuleId);
    }
  }

  // Update project tree
  if (options.projectTree !== undefined) {
    state.projectTree = options.projectTree;
  }

  // Update hint count (DEPRECATED - kept for backward compatibility)
  if (options.hintCount !== undefined) {
    if (!state.hints) {
      state.hints = {};
    }
    state.hints[currentId] = options.hintCount;
  }

  // Update difficulty override
  if (options.difficultyOverride !== undefined) {
    state.difficultyOverride = options.difficultyOverride;
  }

  // Update assessment pending status
  if (options.assessmentPending !== undefined) {
    state.assessmentPending = options.assessmentPending;
  }

  // Add new TODOs to registry
  if (options.addTodos && options.addTodos.length > 0) {
    if (!state.todos) {
      state.todos = {};
    }
    for (const todo of options.addTodos) {
      state.todos[todo.id] = todo;
    }
  }

  // Update existing TODO
  if (options.updateTodo) {
    if (!state.todos) {
      state.todos = {};
    }
    const existing = state.todos[options.updateTodo.id];
    if (existing) {
      state.todos[options.updateTodo.id] = {
        ...existing,
        ...options.updateTodo.updates,
      };
    }
  }

  // Record user mistake
  if (options.addMistake) {
    if (!state.userMistakes) {
      state.userMistakes = [];
    }
    state.userMistakes.push(options.addMistake);
  }

  // Update timestamp
  state.updatedAt = new Date().toISOString();

  // Save updated state
  await saveState(statePath, state);

  return state;
}

/**
 * Get the current module ID from state
 */
export function getCurrentModuleId(state: StateData): string {
  return state.currentModuleId;
}

/**
 * Get the skill score for a module
 */
export function getSkillScore(state: StateData, moduleId: string): number {
  return state.skillScores[moduleId] ?? 0.0;
}

/**
 * Check if a module is completed
 */
export function isModuleCompleted(state: StateData, moduleId: string): boolean {
  return state.completedModules.includes(moduleId);
}

/**
 * Get the hint count for a module
 */
export function getHintCount(state: StateData, moduleId: string): number {
  return state.hints?.[moduleId] ?? 0;
}

/**
 * Increment the skill score for a module
 */
export async function incrementSkillScore(
  statePath: string,
  moduleId: string,
  points: number
): Promise<number> {
  const state = await loadState(statePath);

  const currentScore = state.skillScores[moduleId] ?? 0.0;
  const newScore = currentScore + points;

  // Update the score
  state.skillScores[moduleId] = newScore;
  state.updatedAt = new Date().toISOString();

  // Save updated state
  await saveState(statePath, state);

  return newScore;
}

/**
 * Get module progress for a specific module
 */
export function getModuleProgress(state: StateData, moduleId: string) {
  return (
    state.moduleProgress?.[moduleId] ?? {
      attempts: 0,
      objectivesCompleted: [],
      hintsUsed: 0,
      solutionsUsed: 0,
      bestScore: 0,
    }
  );
}

/**
 * Check if all objectives are completed for a module
 */
export function areAllObjectivesCompleted(
  state: StateData,
  moduleId: string,
  totalObjectives: number
): boolean {
  const progress = getModuleProgress(state, moduleId);
  return progress.objectivesCompleted.length >= totalObjectives;
}

/**
 * Calculate proficiency penalty based on hints and solutions used
 */
export function calculateProficiencyPenalty(hintsUsed: number, solutionsUsed: number): number {
  // Each hint reduces score by 0.5 points
  // Each solution reduces score by 1.5 points
  const hintPenalty = hintsUsed * 0.5;
  const solutionPenalty = solutionsUsed * 1.5;
  return Math.min(hintPenalty + solutionPenalty, 3.0); // Max 3 point penalty
}

/**
 * Get a TODO record by ID
 */
export function getTodoById(state: StateData, todoId: string): import('../types/state.js').TodoRecord | undefined {
  return state.todos?.[todoId];
}

/**
 * Get all TODOs for a specific module
 */
export function getTodosByModule(state: StateData, moduleId: string): import('../types/state.js').TodoRecord[] {
  if (!state.todos) return [];
  return Object.values(state.todos).filter((todo) => todo.moduleId === moduleId);
}

/**
 * Get all pending TODOs
 */
export function getPendingTodos(state: StateData): import('../types/state.js').TodoRecord[] {
  if (!state.todos) return [];
  return Object.values(state.todos).filter((todo) => todo.status === 'pending');
}

/**
 * Get all user mistakes
 */
export function getUserMistakes(state: StateData): import('../types/state.js').UserMistake[] {
  return state.userMistakes || [];
}

/**
 * Get user mistakes for a specific module
 */
export function getMistakesByModule(state: StateData, moduleId: string): import('../types/state.js').UserMistake[] {
  if (!state.userMistakes) return [];
  return state.userMistakes.filter((mistake) => mistake.moduleId === moduleId);
}

/**
 * Get user mistakes for a specific objective
 */
export function getMistakesByObjective(state: StateData, moduleId: string, objectiveIndex: number): import('../types/state.js').UserMistake[] {
  if (!state.userMistakes) return [];
  return state.userMistakes.filter(
    (mistake) => mistake.moduleId === moduleId && mistake.objectiveIndex === objectiveIndex
  );
}

/**
 * Get weakness areas with frequency count
 */
export function getWeaknessAreas(state: StateData, moduleId?: string): Record<string, number> {
  const mistakes = moduleId ? getMistakesByModule(state, moduleId) : getUserMistakes(state);
  const weaknesses: Record<string, number> = {};

  for (const mistake of mistakes) {
    weaknesses[mistake.weaknessArea] = (weaknesses[mistake.weaknessArea] || 0) + 1;
  }

  return weaknesses;
}
